### PR TITLE
Control the resolution of the projector

### PR DESCRIPTION
**Pull request by FinnStutzenstein**
This feature provides the same projectorview and resolution on every device. One can set a resolution for the main projector and the projector scales (if needed) to fit this resolution. The preview in the sidebar stays at a fixed width, so the height of the preview changes with the current resolution.

The old projector is now available under `/real-projector`. The `/projector` shows the real projector in an iframe to control the display and resolution of the real projector.

I've added the fields `width` and `height` to the projector model and added a route `set_resolution`, because with the multiprojector one should be able to set different resolutions per projector. For now, these fields are unused, because I put the resolution in the global settings in fact that OpenSlides do not have a separate `manage-projector` view. But if someone switches to multiprojector, he hasn't to change much of the code; Just delete the config and get the resolution from the `Projector` instead of `Config`.

### CODE REVIEW

File: openslides/core/config_variables.py [hunk 1]
```diff
@@ -155,3 +155,19 @@ def get_config_variables():
...
+        default_value='1024_768',
+        input_type='choice',
+        label='Projector Resolution',
+        choices=(
+            {'value': '800_600', 'display_name': '800x600 (4:3)'},
+            {'value': '1024_768', 'display_name': '1024x768 (4:3)'},
+            {'value': '1280_1024', 'display_name': '1280x1024 (5:4)'},
+            {'value': '1360_768', 'display_name': '1360x768 (16:9)'},
+            {'value': '1680_1050', 'display_name': '1680x1050 (8:5)'},
+            {'value': '1920_1080', 'display_name': '1920x1080 (16:9)'}),
```
Comment (emanuelschuetze) [inline openslides/core/config_variables.py]: It should also bei possible to add customized resolution, e.g. 16:10.

File: openslides/core/static/js/core/projector.js [hunk 2]
```diff
@@ -57,6 +57,60 @@ angular.module('OpenSlidesApp.core.projector', ['OpenSlidesApp.core'])
...
+    'Config',
+    function($scope, Config) {
+        // watch for changes in Config
+        $scope.$watch(function () {
+            return Config.lastModified();
+        }, function () {
+            var split = Config.get('projector_resolution').value.split('_');
+            $scope.projectorWidth = parseInt(split[0]);
+            $scope.projectorHeight = parseInt(split[1]);
+            $scope.recalculateIframe();
```
Comment (emanuelschuetze) [inline openslides/core/static/js/core/projector.js]: Try to recalculate only if the special config is changed - not if any config changed.

File: openslides/core/config_variables.py (hunk 1 above)
Comment (FinnStutzenstein) [inline openslides/core/config_variables.py]: How do you handle custom fields now? Like just adding a file somewhere and read and parse it in the config?

Comment (emanuelschuetze) [inline openslides/core/config_variables.py]: Maybe just a textarea/Input with a Special Input Element like Color picker

**Comment by FinnStutzenstein** [comment]
@emanuelschuetze Do you have some good limits for the resolution? I would say the minimum is 800x600 and maximum FullHD? Or even 4K?

File: openslides/core/urls.py [hunk 3]
```diff
@@ -30,6 +30,9 @@
     # View for the projectors are handelt by angular.
     url(r'^projector.*$', views.ProjectorView.as_view()),
 
+    # Original view without resolutioncontrol for the projectors are handelt by angular.
```
Comment (normanjaeckel) [inline openslides/core/urls.py]: Typo `handled`

File: openslides/core/views.py [hunk 4]
```diff
@@ -315,6 +329,42 @@ def clear_elements(self, request, pk):
...
+        Example:
+
+        {
+            "width": "1024",
+            "height": "768"
+        }
+        """
+        if not isinstance(request.data, dict):
+            raise ValidationError({'detail': 'Data must be a dictionary.'})
+        if (request.data['width'] <= 0 or
```
Comment (normanjaeckel) [inline openslides/core/views.py]: Here you have to validate that someone does not send something like `{"foo": "bar"}`. At the moment you would get a KeyError.

**Comment by normanjaeckel** [comment]
Please add comments to all places where config related changes (resolution field, template, ...) have to be reverted when we change to multiprojector.

**Comment by normanjaeckel** [comment]
@ostcar may have a look at this.

**Comment by ostcar** [comment]
I don't understand the part on the client side. But if it works I am happy :)

**Comment by emanuelschuetze** [comment]
Ok. Looks very good! - @normanjaeckel Ready for last review and merge.

File: openslides/core/static/js/core/projector.js [hunk 5]
```diff
@@ -57,6 +57,64 @@ angular.module('OpenSlidesApp.core.projector', ['OpenSlidesApp.core'])
     }
 ])
 
+// Projector Container Controller
+.controller('ProjectorContainerCtrl', [
+    '$scope',
+    'Config',
+    function($scope, Config) {
+        // watch for changes in Config
+        var last_conf;
```
Comment (normanjaeckel) [inline openslides/core/static/js/core/projector.js]: Where is this last_conf set and used?

File: openslides/core/static/js/core/site.js [hunk 6]
```diff
@@ -1076,6 +1087,21 @@ angular.module('OpenSlidesApp.core.site', [
 
         });
 
+        // watch for changes in Config
+        var last_conf;
```
Comment (normanjaeckel) [inline openslides/core/static/js/core/site.js]: Here the same.

File: openslides/core/static/templates/core/projector-controls.html [hunk 7]
```diff
@@ -9,9 +9,31 @@
         <h4 translate>Live view</h4>
       </a>
       <div uib-collapse="isLiveViewClosed" ng-cloak>
+        <style>
```
Comment (normanjaeckel) [inline openslides/core/static/templates/core/projector-controls.html]: Please indend the inner stuff.

File: openslides/core/static/templates/projector-container.html [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+<title>OpenSlides &ndash; Projector</title>
+<meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=0;">
+<link rel="stylesheet" href="static/css/openslides-libs.css">
+<link rel="stylesheet" href="static/css/projector.css">
+<link rel="icon" href="/static/img/favicon.png">
+<script src="static/js/openslides-libs.js"></script>
+
+<div ng-controller="ProjectorContainerCtrl" class="pContainer">
+  <div>
+    <style>
```
Comment (normanjaeckel) [inline openslides/core/static/templates/projector-container.html]: Here the same.

**Comment by normanjaeckel** [comment]
Some stylistic remarks.

**Comment by FinnStutzenstein** [comment]
@normanjaeckel @emanuelschuetze ping

**Pull request merged**